### ISSUE TITLE
Godot crashes immediately when opening a project with "RtAudio-DirectSound" set as audio driver

### ISSUE THREAD

**Issue by wingedadventurer** [godotengine/godot#25503]
**Godot version:**
3.1 beta 3 (though it started to crash around 3.1 alpha 4 too I think)

**OS/device including version:**
Windows 10

**Issue description:**
Check minimal project. Engine crashes without any warning messages.
My game with "RtAudio-DirectSound" set as audio driver started crashing since 3.1 alpha 4 I think, but I'm not sure if that was the initial cause back then. I've just discovered that audio driver is the cause in 3.1 beta 3.

**Steps to reproduce:**
Start new project. Set audio driver to "RtAudio-DirectSound" in project settings. Save and try to open the project afterwards.

**Minimal reproduction project:**
[RtAudio-DirectSound crash.zip](https://github.com/godotengine/godot/files/2816285/RtAudio-DirectSound.crash.zip)

**Comment by marcelofg55** [comment]
I just tested the project with latest master and it didn't crash for me. Can you try with latest master too?

**Comment by wingedadventurer** [comment]
It still crashes on my end with latest daily build from [here](https://hugo.pro/projects/godot-builds/).

Here's some additional info:

I checked and it crashes since alpha 4. Alpha 3 and previous versions work fine.

I'm using an HP laptop. My audio drivers are up-to-date:

![02-08-2019 02-48-45 pm](https://user-images.githubusercontent.com/26509794/52482151-cb815500-2bb0-11e9-90b5-f5d58efa040f.png)
![02-08-2019 02-48-51 pm](https://user-images.githubusercontent.com/26509794/52482153-cb815500-2bb0-11e9-8c9d-91d43353ba0a.png)

If it's any relevant, DTS Studio Sound is constantly running on my laptop after drivers have installed:

![02-08-2019 02-49-07 pm](https://user-images.githubusercontent.com/26509794/52482190-ece24100-2bb0-11e9-968c-1012e1bf1063.png)

I'm not using any external audio device at the moment, just the speakers.

@marcelofg55

**Comment by marcelofg55** [comment]
Any chance that you could run the project through Visual Studio or WinDBG perhaps to see where it's crashing?

**Comment by wingedadventurer** [comment]
I haven't done such debugging before, could you point me to the documentation on how to set it up?

**Comment by marcelofg55** [comment]
You can compile godot with `scons p=windows vsproj=yes` to create a solution (it creates a godot.sln file) that you can open with Visual Studio. Using that solution you can run godot with Visual Studio, it should catch the crash.

**Comment by wingedadventurer** [comment]
I'll get on it in a day or two, thanks for pointers.

**Comment by wingedadventurer** [comment]
@marcelofg55 Sincere apologies for the delay.

So funny thing happens.

I installed Python v3.7.0, pywin, and Scons v3.0.4. I had latest VS2017 installed already, I just added the C++ module to it.

I cloned Godot source and compiled it successfully for Visual Studio using "scons -j8 p=windows vsproj=yes" in "x64 Native Tools Command Prompt for VS 2017" console.

I opened the solution in VS2017 and ran the solution successfully.

I created new project, set audio driver to "RtAudio-DirectSound", restarted the engine and everything worked fine, no crashes or errors at all.

Then I downloaded latest daily build from [here](https://hugo.pro/projects/godot-builds/) again (64-bit editor, ZIP archive). Unpacked, ran the same project I just created and it crashed without errors like the initial problem.

¯\\\_(ツ)\_/¯

Maybe there is something specific with the way Hugo Locurcio (Calinou) compiles it. I could try asking him and replicate his procedure as precise as possible. Any other ideas?

**Comment by wingedadventurer** [comment]
Also if it's relevant, the builds I compared are:

* Calinou's daily build version: v3.1.beta.calinou.6607c3c
* My build version: v3.1.beta.custom_build.fa7329f

I've checked the commits between them and don't see any significant changes on audio.

**Comment by marcelofg55** [comment]
Maybe Calinou compiles it with mingw?

**Comment by akien-mga** [comment]
Might be related to MinGW and/or LTO yeah, cc @hpvb.

**Comment by hpvb** [comment]
Weren't we going to deprecate rtaudio?

**Comment by akien-mga** [comment]
That would be a solution, yeah :)
I mentioned it to @marcelofg55 a few weeks ago and the one thing that might require keeping RtAudio is if we decide to implement #24274. But we could always bring it back at that point if we do implement it.

That being said, if OP had specific reasons to use it, that means that removing RtAudio support might be a breaking change for some users.

**Comment by wingedadventurer** [comment]
The reason I'm using RtAudio is that it provides me best results. Since I'm making a rhythmical game, precise music playback is essential. Default WASAPI driver for Windows doesn't offer me such results.

I've checked the audio drivers I have in editor (printing `OS.get_audio_name(i)`) and all I get are:

1. WASAPI - default one, terrible latency
2. RtAudio-DirectSound - acceptable latency, but causes crash as explained
3. Dummy - not sure if this is really an audio driver, but it doesn't play any sound for me

Therefore RtAudio is my only option.



Also, an update on testing different compile methods. Seems like compiling with `use_mingw=yes` doesn't cause the issue as well. Now I'm trying to compile with `use_lto=yes` but getting some errors so it might take a while. Also very recent 3.1 beta 4 crashes as well.

**Comment by marcelofg55** [comment]
@wingedadventurer WASAPI should offer better latency actually, what's your audio setup? What Audio Card you're using? And what sample rate are you using?

**Comment by wingedadventurer** [comment]
The drivers I have are listed at the beginning of this discussion (Realtek High Definition Audio). I've checked for driver updates, none are found.

WASAPI doesn't offer me as much settings as RtAudio does. These are the options I have in project settings:

![02-12-2019 07-49-54 pm](https://user-images.githubusercontent.com/26509794/52660263-a1f16200-2eff-11e9-8448-2b51c77449f2.png)

Basically in my game I'm playing the beats, using AnimationPlayer as a sequencer. When I measured time between equally spaced beats, there is no latency in code, only in audio. Latency is greatly reduced with RtAudio so it's highly likely due to the audio driver. The beats that I play are of type .wav and audio import settings are left default:

![02-12-2019 07-56-18 pm](https://user-images.githubusercontent.com/26509794/52660594-4d021b80-2f00-11e9-99be-acb9833d7bff.png)

**Comment by marcelofg55** [comment]
@wingedadventurer Are you using an internal Audio Card? or an external one? And the sample rate?
You can check the sample rate in windows on the properties of the audio device:
![audioconfiguration](https://user-images.githubusercontent.com/10578225/52662147-61372000-2ee2-11e9-84f3-cf871ac91c9b.png)

**Comment by wingedadventurer** [comment]
@marcelofg55 They're default settings, I never modify sound properties. I think it's an internal audio card (it's a laptop).

![02-12-2019 08-25-50 pm](https://user-images.githubusercontent.com/26509794/52662416-6efd9d00-2f04-11e9-95e4-404007b6ee4b.png)

But I think I found the problem. 😄 Compiling the source on Windows. with `use_lto=yes` makes the engine crash. Compiling without LTO doesn't produce crash.

**Comment by marcelofg55** [comment]
I see, I'll check the crash using `use_lto=yes` then, thanks for finding that!

Can you also try switching the sample rate to 44100 Hz? It should give a much better latency with that using the WASAPI driver, please let me know if that really helps.

**Comment by wingedadventurer** [comment]
@marcelofg55 It worked! Setting sample rate to 44100 Hz (or amazingly anything else other than 48000 Hz I had) fixed the latency! 😃 I'm just reading on sample rates and only now find that the standard rate is 44 KHz. I'll yet have to educate myself better on that field. Good to know where to look for future potential audio errors.

Thank you very much, you saved me a lot of trouble! Kudos!

Joy aside, while my latency problem is solved, the RtAudio one still isn't so I will assist further into debugging it now that I have some source compiling experience and the tools (and I would love to make at least one PR for one of the most awesome engines out there).

I assume I have to enable debugging somehow for the last command I used to compile:
```
scons -j1 platform=windows target=release_debug use_mingw=yes use_lto=yes
```
How can I do that?

**Comment by marcelofg55** [comment]
Glad it's working better now!

`target=release_debug` already has debugging support, so you should be able to debug with that build, though it doesn't create a solution for you that way (`vsproj=yes` does that).

**Comment by wingedadventurer** [comment]
@marcelofg55 I compiled with `scons -j1 p=windows vsproj=yes use_lto=yes` and got the VS project, but the crash doesn't appear. I assume it's the `use_mingw` and `use_lto` combination that causes the crash.

I'm unable to compile with `vsproj=yes` and `use_mingw=yes` flags at the same time, whether I use default prompt or the VS one.

**Comment by marcelofg55** [comment]
@wingedadventurer You might have to use an external debugger then, maybe WinDBG. I don't have much experience debugging mingw issues so maybe someone with mingw knowledge can help out.

**Comment by marcelofg55** [comment]
I could finally debug this issues and found that it's crashing on the function `deviceQueryCallback` at RtAudio.cpp

**Comment by wingedadventurer** [comment]
I'm not having much luck with successfully compiling the engine with debug symbols to use with WinDBG after several attempts and some research so I can't really help in that part.

**Comment by marcelofg55** [comment]
It's crashing on this line: https://github.com/godotengine/godot/blob/9f57939c121a76280d152ec1e6a2c2d3bd9dc169/thirdparty/rtaudio/RtAudio.cpp#L6718
If I comment that line it crashes a bit later in that func, here:
https://github.com/godotengine/godot/blob/9f57939c121a76280d152ec1e6a2c2d3bd9dc169/thirdparty/rtaudio/RtAudio.cpp#L6736
There must be some bug with `-flto` and the use of `std::string`.

**Comment by akien-mga** [comment]
>  There must be some bug with `-flto` and the use of `std::string`.

@hpvb mentioned to me earlier that there is a known issue of MinGW where using `-flto` and `<iostream>` can cause issues. Removing all the `std::cout` prints might fix it, but another alternative is to just drop RtAudio (#26086).

**Comment by starry-abyss** [comment]
How comes though that for @wingedadventurer WASAPI has bad defaults in Windows? Does it mean forcing players to manually adjust them? Or it's possible to force-change them from Godot?

**Comment by akien-mga** [comment]
> How comes though that for @wingedadventurer WASAPI has bad defaults in Windows? Does it mean forcing players to manually adjust them? Or it's possible to force-change them from Godot?

I'm not sure, but that should probably be discussed in a new issue, so that it can be fixed eventually.

**Comment by wingedadventurer** [comment]
@starry-abyss Non-default settings probably have something to do with me modifying sound settings once long ago, way before starting on my project, and I might have accidentally changed the bit rate.

**Comment by starry-abyss** [comment]
OK then.

**Comment by starry-abyss** [comment]
Hopefully, it doesn't lead to a situation where musicians have to revert their music production settings to defaults each time to be able to play a music game...

**Comment by wingedadventurer** [comment]
What could help is being able to detect current system bitrate settings (and with that maybe some additional useful settings) so the game could at least warn the player that he won't be getting the best experience.
